src/diskusage.cpp

```diff
@@ -0,0 +1,176 @@
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
+ */
+
+#include "diskusage.h"
+#include "diskusage_p.h"
+
+#include <QThread>
+#include <QDir>
+#include <QProcess>
+
+DiskUsageWorker::DiskUsageWorker(QObject *parent)
+    : QObject(parent)
+{
+}
+
+DiskUsageWorker::~DiskUsageWorker()
+{
+}
+
+static quint64 calculateSize(const QString &directory)
+{
+    if (!QDir(directory).exists()) {
+        return 0L;
+    }
+
+    // TODO: Recursively sum up disk usage manually while avoiding symlinks
+    // instead of exec'ing du for every app
+
+    QProcess du;
+    du.start("du", QStringList() << "-sb" << directory, QIODevice::ReadOnly);
+    du.waitForFinished();
+    QStringList size_directory = QString::fromUtf8(du.readAll()).split('\t');
+
+    if (size_directory.size() > 1) {
+        return size_directory[0].toULongLong();
+    }
+
+    return 0L;
+}
+
+void DiskUsageWorker::submit(QStringList paths)
+{
+    foreach (const QString &path, paths) {
+        // TODO: Detect and handle errors (emit error(path, "..."))
+        emit calculated(path, calculateSize(path));
+    }
+
+    emit finished();
+}
+
+
+class DiskUsagePrivate
+{
+    Q_DISABLE_COPY(DiskUsagePrivate)
+    Q_DECLARE_PUBLIC(DiskUsage)
+
+    DiskUsage * const q_ptr;
+
+public:
+    DiskUsagePrivate(DiskUsage *usage);
+    ~DiskUsagePrivate();
+
+private:
+    QThread m_thread;
+    DiskUsageWorker *m_worker;
+};
+
+DiskUsagePrivate::DiskUsagePrivate(DiskUsage *usage)
+    : q_ptr(usage)
+    , m_thread()
+    , m_worker(new DiskUsageWorker())
+{
+    m_worker->moveToThread(&m_thread);
+
+    QObject::connect(usage, SIGNAL(calculateStart(QStringList)),
+                     m_worker, SLOT(submit(QStringList)));
+
+    QObject::connect(m_worker, SIGNAL(error(QString, QString)),
+                     usage, SIGNAL(error(QString, QString)));
+
+    QObject::connect(m_worker, SIGNAL(calculated(QString, quint64)),
+                     usage, SIGNAL(calculated(QString, quint64)));
+
+    QObject::connect(m_worker, SIGNAL(calculated(QString, quint64)),
+                     usage, SLOT(calculateResult(QString, quint64)));
+
+    QObject::connect(m_worker, SIGNAL(finished()),
+                     usage, SLOT(calculateFinished()));
+
+    m_thread.start();
+}
+
+DiskUsagePrivate::~DiskUsagePrivate()
+{
+    m_worker->deleteLater();
```

Comment:
The worker thread will remain in its exec() waiting for events unless it is stopped. Destroying a QThread instance does not stop the underlying system thread => resource leak.